tools/libInMemoryFrontend/include/libInMemoryFrontend/InMemoryFrontend.h

```diff
@@ -0,0 +1,41 @@
+//===--- InMemoryFrontend.h - Frontend operations, in memory ----*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_LIBINMEMORYFRONTEND_INMEMORYFRONTEND_H
+#define SWIFT_LIBINMEMORYFRONTEND_INMEMORYFRONTEND_H
+
+#include "swift/Frontend/Frontend.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+namespace swift {
+namespace inmemoryfrontend {
+
+/// Given a fully setup CompilerInstance, configured to emit one module, runs
+/// the compilation and emits the module to a memory buffer, without writing to
+/// the filesystem. Emits error information to the CompilerInstance's
+/// DiagnosticEngine.
+///
+/// \param moduleBuffer will be set to a pointer to the serialized module
+///                     buffer. nullptr is allowed, in which case the module
+///                     will not be serialized.
+/// \param moduleDocBuffer will be set to a pointer to the serialized module
+///                        doc buffer. nullptr is allowed, in which case the
+///                        module doc will not be serialized.
+/// \return true on error.
+bool compileSwiftModule(CompilerInstance &CI,
+                        std::unique_ptr<llvm::MemoryBuffer> *moduleBuffer,
```

Comment:
Hm, we do this enough it might be worth it to make a move-only type to encapsulate this pattern...

Reply:
Which pattern? Having pointer for an output parameter, where the caller can set it to `nullptr` when the caller isn't interested in that output?

Reply:
Yeah, passing around a package of nullable input buffers. It's just an idle thought. I wouldn't block this on it.